### PR TITLE
[TTAHUB-1892] Admin TR import page

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

This is still a WIP as we don't have the BE yet.

As an admin visit the admin page there is a new button called 'Training Reports'.

## How to test

**Note:** Will have to mock the BE response from fetchers\admin.js

Test the following cases:
- Throws a error if the user imports a file not of CSV type.
- Throws an error if missing required columns.
- Throws an error if we have invalid columns.
- Throws and error if file contains duplicate event Id's.
- Successfully imports a good file (letting the user know if events had errors or skipped).

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1892

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/pages/Admin/TrainingReports.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            </Alert>
+          )}
+          {info && (
+          <Alert type="info" className="margin-bottom-1 maxw-mobile-lg" noIcon>
+            {info}
+          </Alert>
+          )}
+          <div className="display-flex">
+            <FormGroup>
+              <Label htmlFor="file-input-single">Input accepts a single file</Label>
```
Comment (thewatermethod) [inline frontend/src/pages/Admin/TrainingReports.js:161]: htmlFor has to match the id on the FileInput

File: frontend/src/pages/Admin/TrainingReports.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+              {(success) && (
+              <div>
+                <h3>Import Summary:</h3>
+                <ul>
+                  <li>{`${skipped.length} skipped: ${skipped.join(', ')} `}</li>
+                  <li>{`${errors.length} errors: ${errors.join(', ')}`}</li>
+                </ul>
+              </div>
+              )}
+              <Button className="margin-top-2" type="button" onClick={importTr} disabled={uploadDisabled}>Upload training reports</Button>
```
Comment (thewatermethod) [inline frontend/src/pages/Admin/TrainingReports.js:172]: Would prefer not to disable button and just bomb out of the "importTr" function if there is an error showing

**Review by thewatermethod — COMMENTED** [review]
Looks great, just a couple of comments

File: frontend/src/pages/Admin/TrainingReports.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+      // Check if we have duplicate 'Event ID' values.
+      if (distinctEventIds.length !== eventIds.length) {
+        setError('Duplicate Event IDs found. Please correct and try again.');
+        setUploadDisabled(true);
+        e.target.value = null;
+        return;
+      }
+
+      // Display count of distinct ids.
+      setInfo(`${distinctEventIds.length} events will be imported.`);
```
Comment (nvms) [inline frontend/src/pages/Admin/TrainingReports.js:129]: If the file only has headers and no rows, you'll see this:

Screenshot 2023-09-07 at 3 11 53 PM

Not really a big deal or bug IMO, but thought I'd mention it.

File: frontend/src/pages/Admin/index.js [hunk 4]
```diff
@@ -40,6 +41,9 @@ function Admin() {
         <NavLink activeClassName="usa-button--active" className="usa-button" to="/admin/email">
           Email
         </NavLink>
+        <NavLink activeClassName="usa-button--active" className="usa-button" to="/admin/training-reports">
```
Comment (nvms) [inline frontend/src/pages/Admin/index.js:44]: Should this be behind the same feature flag used below?

https://github.com/HHS/Head-Start-TTADP/blob/383fd87c6791aa191783cf61c4d02bb8112073fd/frontend/src/App.js#L266-L270

**Review by nvms — COMMENTED** [review]
Looks great so far. Not ready to approve just yet - have some comments & questions.

**Comment by nvms** [comment]
Also, I tested all these:

> Throws a error if the user imports a file not of CSV type.
Throws an error if missing required columns.
Throws an error if we have invalid columns.
Throws and error if file contains duplicate event Id's.

Works great and I think these cases were handled very nicely. 👍

File: frontend/src/pages/Admin/TrainingReports.js (hunk 3 above)
Comment (nvms) [reply frontend/src/pages/Admin/TrainingReports.js:129]: Thought about it - not really worth commenting on this.

File: frontend/src/pages/Admin/TrainingReports.js (hunk 1 above)
Comment (AdamAdHocTeam) [reply frontend/src/pages/Admin/TrainingReports.js]: Done.

File: frontend/src/pages/Admin/TrainingReports.js (hunk 3 above)
Comment (AdamAdHocTeam) [reply frontend/src/pages/Admin/TrainingReports.js:129]: I can add a check.

Comment (AdamAdHocTeam) [reply frontend/src/pages/Admin/TrainingReports.js:127]: Per discussion will leave as is for now.

**Review by thewatermethod — CHANGES_REQUESTED** [review]
I'm thinking you haven't seen my comment from last week, one change still to make

File: frontend/src/pages/Admin/TrainingReports.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      setErrors([]);
+
+      // Get the file from the file input from ref.
+      const { files } = fileInputRef.current;
+      const file = files[0];
+
+      // Get File.
+      const data = new FormData();
+      data.append('file', file);
+      const res = await importTrainingReports(file);
```
[the remark this answers is not in the document]
Comment (AdamAdHocTeam) [reply frontend/src/pages/Admin/TrainingReports.js]: done.

**Review by thewatermethod — APPROVED** [review]

File: src/services/event.ts [hunk 6]
```diff
@@ -444,3 +445,112 @@ export async function findAllEvents(): Promise<EventShape[]> {
...
+  'Event ID': 'eventId',
+  'Overall Vision/Goal for the PD Event': 'vision',
+  'Reason for Activity': 'reasons',
+  'Target Population(s)': 'targetPopulations',
+  'Event Organizer - Type of Event': 'eventOrganizer',
+};
+
+const toSplit = ['targetPopulations', 'reasons'];
+
+const replacements: Record<string, string> = {
```
Comment (AdamAdHocTeam) [inline src/services/event.ts:465]: Do we want to add some tests for the auto replacement stuff?

File: src/services/event.ts [hunk 7]
```diff
@@ -444,3 +445,112 @@ export async function findAllEvents(): Promise<EventShape[]> {
...
+    data[mappedKey] = toSplit.includes(mappedKey)
+      ? splitPipe(line[key]).map(applyReplacements)
+      : line[key];
+  });
+
+  return data;
+};
+
+const checkUserExists = async (creator: string) => {
+  const user = await db.User.findOne({ where: { email: creator } });
```
Comment (AdamAdHocTeam) [inline src/services/event.ts:487]: Do we want to verify the user can Read/Write training reports?

File: src/services/event.ts [hunk 8]
```diff
@@ -444,3 +445,112 @@ export async function findAllEvents(): Promise<EventShape[]> {
...
+  'Edit Title': 'eventName',
+  'Event Duration/#NC Days of Support': 'eventDuration',
+  'Event ID': 'eventId',
+  'Overall Vision/Goal for the PD Event': 'vision',
+  'Reason for Activity': 'reasons',
+  'Target Population(s)': 'targetPopulations',
+  'Event Organizer - Type of Event': 'eventOrganizer',
+};
+
+const toSplit = ['targetPopulations', 'reasons'];
```
Comment (AdamAdHocTeam) [inline src/services/event.ts:463]: Do we assert the new array created in a test?

File: src/services/event.ts (hunk 7 above)
Comment (nvms) [reply src/services/event.ts:487]: This is a good question to pose to a wider audience - maybe @ standup

Comment (thewatermethod) [reply src/services/event.ts:487]: We by convention always check permissions in the handlers, not in the services. If this is called in the admin namespace, then a user will be authenticated as an admin before calling this.

Comment (thewatermethod) [reply src/services/event.ts:487]: deleted my comment as it was dumb

**Review by AdamAdHocTeam — COMMENTED** [review]
As we discussed:
- We would want a test for this service function
- Also we don't want to have similar code  in two places (cmd import and here)
- Probably can ask Angela if we can remove the cmd import

File: src/services/event.ts (hunk 6 above)
Comment (AdamAdHocTeam) [reply src/services/event.ts:465]: This meant, add a row to the unit testing file for this use case.

File: src/services/event.ts (hunk 8 above)
Comment (AdamAdHocTeam) [reply src/services/event.ts:463]: Again this just means to validate the resulting array in a unit test.

**Review by AdamAdHocTeam — COMMENTED** [review]
Did some testing overall works great. I would just cleanup the old code and update the test.

Also raise the question about validating user permissions.

File: src/services/event.ts (hunk 8 above)
Comment (nvms) [reply src/services/event.ts]: The tests now check for this

File: src/services/event.ts (hunk 7 above)
Comment (nvms) [reply src/services/event.ts]: We confirmed in standup that we need to not import if the user defined in the line item doesn't have write permissions. This is complete and tests are written for it

File: src/services/event.ts (hunk 6 above)
Comment (nvms) [reply src/services/event.ts:470]: The tests account for this now

**Comment by nvms** [comment]
@AdamAdHocTeam I think I addressed all the comments. I also removed the importSmartSheetEvent stuff

File: src/routes/admin/trainingReport.js [hunk 9]
```diff
@@ -19,6 +19,12 @@ export async function importTrainingReport(req, res) {
         return;
       }
 
+      // allow text/csv only
+      if (files.file[0].headers['content-type'] !== 'text/csv') {
```
Comment (AdamAdHocTeam) [inline src/routes/admin/trainingReport.js:23]: Might want to add a test for this path.

**Comment by thewatermethod** [comment]
Is there a file we can use to test this?

**Comment by nvms** [comment]
> Is there a file we can use to test this?

Yep! There's one attached to the ticket (https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1892). It's the one named "Prod TR Events Import_...."

**Comment by thewatermethod** [comment]
> Yep! There's one attached to the ticket

D'oh.

Works good - we probably just want to fix the failing test, as it relates to this change

**Pull request merged**